proposals/0030-markdoc.md

```diff
@@ -0,0 +1,295 @@
+- Future: **Full alignment with Markdown and MDX** rendered result. Namely, the computed `headings` property (which can be tackled in future releases) and frontmatter manipulation via remark (since remark is incompatible with Markdoc).
+
+# Detailed Design
+
+## API
+
+Markdoc support will be specific to content collections. This means Markdoc files can be included alongside other content collection entries via the `.mdoc` extension. Entries can also be queried using the same `getCollection()` and `getEntryBySlug()` APIs for Markdown and MDX.
+
+### Rendering Markdoc content
+
+Users can render contents and configure Markdoc via the `<Content />` component. This will be exposed from the `render()` result, and feature two props:
+
+- `config?: import('@markdoc/markdoc').Config`: An (optional) Markdoc config to be used [during the transformation step](https://markdoc.dev/docs/render#transform). This includes configuration for Markdoc tags and variables.
+- `components?: Record<string, ComponentRenderer>`: An (optional) mapping from Markdoc tags or elements to Astro components.
+
+We expect users to share a common Markdoc `config` and `components` setup throughout their application. Because of this, we will recommend creating a utility component to encapsulate that config.
+
+For example, say a user wants to share config specific to their `blog` collection. This `BlogContent` component will add support for an `{% aside %}` tag across blog entries:
+
+```astro
+---
+// src/components/BlogContent.astro
+import Aside from './Aside.astro';
+import type { CollectionEntry } from 'astro:content';
+
+type Props = {
+	entry: CollectionEntry<'blog'>;
+};
+
+const { entry } = Astro.props;
+const { Content } = await entry.render();
+---
+
+<Content
+	config={{
+		tags: {
+			aside: {
+				render: 'Aside',
+				attributes: {
+					type: { type: String },
+					title: { type: String },
+				},
+			},
+		},
+	}}
+	components={{ Aside }}
+/>
+```
+
+Now, users can render any blog collection entry with this config by passing the entry to `BlogContent`:
+
+```astro
+---
+import { getEntryBySlug } from 'astro:content';
+import BlogContent from '../components/BlogContent.astro';
+
+const mdocEntry = await getEntryBySlug('blog', 'test');
+---
+
+<h1>{intro.data.title}</h1>
+<BlogContent entry={mdocEntry} />
+```
+
+### Advanced: mapping Markdoc nodes to component props
+
+Component renderers can also include a `props()` function to map Markdoc attributes and AST entries to component props. This is useful when:
+
+- computing props based on the Markdoc AST
+- mapping Markdoc's generated attributes to prop names
+
+This example maps Markdoc's generated `data-language` attribute for code blocks to the `lang` prop used by Astro's `Code` component, and stringifies the contents to HTML for use with Shiki:
```

Comment:
Do we have any other use-cases? If this is the only one, then I'd say this is out of scope and the `Code` component could just support this, to be revisited when we had a real use-case.

If we already have other/multiple use-cases, then okay to ignore this!

Reply:
This was the initial use case that prompted the need. I'll come back to you if there are more clear use cases. Otherwise, we can remove!

Reply:
Update: realized Astro's `slots.render()` can fill this use case well. Removing from RFC and the final API!